doc/design/python_interface.md

```diff
@@ -0,0 +1,99 @@
+# Python Interface Design
+
+## Variable as variable data type (not parameter)
+The class Variable is introduced to define the variables, which are not only parameters but also the input data and intermediate outputs.
+
+All the variables are defined in a variable scope, `Scope`, it will check duplicate variable names.
+
+The `Variable`'s definition is as follows
+```python
+class UniqueNameGenerator:
+    counter = 0
+
+    def __call__(self, prefix):
+        name = '%s-%d' % (prefix, counter)
+        counter += 1
+        return name
+
+
+unique_name_generator = UniqueNameGenerator()
+
+
+class Variable(object):
+    def __init__(self, shape, name=None, prefix=None, var_scope=None, op=None):
+        '''
+        var_scope is VarDesc's scope.
+        '''
+        if name is None:
+            if prefix is not None:
+                name = unique_name_generator(prefix)
```

Comment:
I like this design. We don't need to call `unique_name_generator` in layer definitions.